pennylane/_qubit_device.py

```diff
@@ -688,3 +689,113 @@ def sample(self, observable):
+
+    def adjoint_jacobian(self, tape):
+        """Implements the method outlined in https://arxiv.org/abs/2009.02823 to calculate the
+        Jacobian."""
+
+        for m in tape.measurements:
+            if m.return_type is not qml.operation.Expectation:
+                raise qml.QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+            if not hasattr(m.obs, "base_name"):
+                m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        # Perform the forward pass
+        self.reset()
+
+        # Consider using caching and calling lower-level functionality. We just need the state
+        # without postprocessing https://github.com/PennyLaneAI/pennylane/pull/1032/files#r563441040
+        self.execute(tape)
+
+        phi = self._reshape(self.state, [2] * self.num_wires)
+
+        lambdas = [self._apply_operation(phi, obs) for obs in tape.observables]
+
+        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+
+        expanded_ops = []
+        for op in reversed(tape.operations):
+            if op.num_params > 1:
+                if isinstance(op, qml.Rot) and not op.inverse:
+                    ops = op.decomposition(*op.parameters, wires=op.wires)
+                    expanded_ops.extend(reversed(ops))
+                else:
+                    raise QuantumFunctionError(
+                        f"The {op.name} operation is not supported using "
+                        'the "adjoint" differentiation method'
+                    )
+            else:
+                if not op.name in ("QubitStateVector", "BasisState"):
+                    expanded_ops.append(op)
+
+        dot_product_real = lambda a, b: self._real(math.sum(self._conj(a) * b))
+
+        param_number = len(tape._par_info) - 1  # pylint: disable=protected-access
+        trainable_param_number = len(tape.trainable_params) - 1
+        for op in expanded_ops:
+
+            if (op.grad_method is not None) and (param_number in tape.trainable_params):
+                d_op_matrix = operation_derivative(op)
+
+            op.inv()
+            phi = self._apply_operation(phi, op)
+
+            if op.grad_method is not None:
+                if param_number in tape.trainable_params:
+                    mu = self._apply_unitary(phi, d_op_matrix, op.wires)
+
+                    jac_column = np.array(
+                        [2 * dot_product_real(lambda_, mu) for lambda_ in lambdas]
+                    )
+                    jac[:, trainable_param_number] = jac_column
+                    trainable_param_number -= 1
+                param_number -= 1
+
+            lambdas = [self._apply_operation(lambda_, op) for lambda_ in lambdas]
+            op.inv()
+
+        return jac
+
+
+def operation_derivative(operation) -> np.ndarray:
```

Comment:
This function seems to belong rather into `operations.py`

Reply:
Good point, it's now been moved there: e0a9d85504b6d1a72afda4803dfba130722a98ef